lib/algos/list/delimiter.js

```diff
@@ -70,67 +88,72 @@ class Delimiter {
         });
         this.NextMarker = key;
         ++this.keys;
+        return this;
     }
 
     /**
-     *  This function add a common prefix to the CommonPrefixes
-     *  Set the NextMarker to the current commonPrefix
-     *  Increment the keys counter
-     *  @param {String} commonPrefix - The commonPrefix to process
-     *  @return {undefined}
-     */
-    addCommonPrefix(commonPrefix) {
-        if (this.CommonPrefixes.indexOf(commonPrefix) === -1) {
-            this.CommonPrefixes.push(commonPrefix);
-            this.NextMarker = commonPrefix;
-            ++this.keys;
-        }
-    }
-
-    /**
-     *  This function apply the filter to each element
-     *  It's looking for the next delimiter
-     *  if nextDelim === -1 can mean two things
-     *       -> this.delimiter is not present in the key after the prefix
-     *  else
-     *       -> the key is `${this.prefix}${this.delimiter}${objName}`
-     *  @param {Object} obj - The key and value of the element
-     *  @param {String} obj.key - The key and value of the element
-     *  @param {String} obj.value - The key and value of the element
-     *  @return {Boolean} - True: Continue, False: Stop
+     *  Filter to apply on each iteration, based on:
+     *  - prefix
+     *  - delimiter
+     *  - maxKeys
+     *  The marker is being handled directly by levelDB
+     *  @param {Object} obj       - The key and value of the element
+     *  @param {String} obj.key   - The key of the element
+     *  @param {String} obj.value - The value of the element
+     *  @return {Boolean}         - indicates if iteration should continue
      */
     filter(obj) {
-        // Check first in case of maxkeys <= 0
+        const key = obj.key;
+        const value = obj.value;
         if (this.keys >= this.maxKeys) {
-            // In cases of maxKeys <= 0 => IsTruncated = false
+            // In cases of maxKeys <= 0 -> IsTruncated = false
             this.IsTruncated = this.maxKeys > 0;
             return false;
         }
-        const key = obj.key;
-        const value = obj.value;
-        if (this.prefix && !key.startsWith(this.prefix)) {
+        if (this.prefix && !key.startsWith(this.prefix)
+                && key.startsWith(this.NextMarker)) {
```

Comment:
Which highlights even more the lack of testing outside of the happy path then. Changing to `key === this.NextMarker` should do the trick.